src/multi.c

```diff
@@ -354,36 +415,47 @@ int isWatchedKeyExpired(client *c) {
 
 /* "Touch" a key, so that if this key is being WATCHed by some client the
  * next EXEC will fail. */
+/* 如果 key 正在被其他客户端监视，执行 exec 将会失败 */
 void touchWatchedKey(redisDb *db, robj *key) {
     list *clients;
     listIter li;
     listNode *ln;
 
+    /* 如果数据库中没有被监视的 key，直接范湖 */
     if (dictSize(db->watched_keys) == 0) return;
+    /* 获取监视给定 key 的客户端列表 */
     clients = dictFetchValue(db->watched_keys, key);
+    /* 没有客户端监视该 key，返回 */
     if (!clients) return;
 
     /* Mark all the clients watching this key as CLIENT_DIRTY_CAS */
     /* Check if we are already watching for this key */
+    /* 遍历所有监视该 key 的客户端 */
     listRewind(clients,&li);
     while((ln = listNext(&li))) {
         watchedKey *wk = listNodeValue(ln);
         client *c = wk->client;
 
+        /* 如果客户端的 watched_key 过期了 */
         if (wk->expired) {
             /* The key was already expired when WATCH was called. */
+            /* 判断是否 watched_key 对应的 key 是否在数据库中已经被清除了
+             * 如果清除了，可以清除该 watched_key 的过期标识 */
             if (db == wk->db &&
                 equalStringObjects(key, wk->key) &&
                 dictFind(db->dict, key->ptr) == NULL)
             {
                 /* Already expired key is deleted, so logically no change. Clear
                  * the flag. Deleted keys are not flagged as expired. */
                 wk->expired = 0;
+                /* 这里会使用 continue 将所有监视该 key 的客户端的对应的 watched_key 的过期标识进行修改 */
                 goto skip_client;
             }
+            /* 如果 watched_key 过期，但是对应的 key 还存在，直接退出当前循环 */
```

Comment:
这个我其实最早到现在，一直没想到有啥场景会走到 break，你有想到吗
```suggestion
            /* 如果 watched_key 逻辑过期，但是对应的 key 还存在，直接退出当前循环 */
```

Reply:
这个函数的调用时机是做了修改key的动作就会调用该函数，但是修改key的地方并不一定都会做过期删除逻辑。
而判断key是否过期，使用了server.fixed_time_expire：
如果这个值大于0的情况，用来判断是否过期的当前时间是redis缓存的时间戳，可以认为这个属性大于0的情况，更新时间戳后也一定会调用databaseCron中删除过期键的逻辑，也就是一定不会到break；
如果这个属性值是不大于0的情况，是直接取操作系统的时间戳作为当前时间，应该会存在走到break这个地方的情况

Reply:
哦哦哦 thanks, 说到 fixed_time_expire 我就回想起来了，我忘记了逻辑过期的键不一定每次被打开时都会被删除

btw, fixed_time_expire 在 7.2 里被干掉了